### DISCUSSION TITLE
Component fields are missing in API response

### DISCUSSION THREAD

**Discussion by s-barsch** [strapi/strapi#5769] [Legacy Topics]
**Describe the bug**

Component fields are missing in API response (if their types are referenced by a relation).

**Steps to reproduce the behavior**

1. Create `piece` type:
```
title Text
info Component
    material Text
    color Text
```
2. Create a single `selection` type:
```
entry Component (repeatable)
    piece Relation with *piece* (Selection has one Piece)
```
3. Create pieces and a selection of these pieces.
4. Go to `http://localhost:1337/selection`.

**Expected behavior**

I expect to get a list of pieces on `/selection`. A selection is returned, but the pieces have no `info` field. Only their title is returned. `/pieces`, on the other hand, works fine and as expected, with the `info` field included in the response.

**System**
- Node.js version: v12.16.2
- NPM version: 6.14.4 
- Strapi version: 3.0.0-beta.19.5
- Database: Sqlite
- Operating system: Debian

**Comment by lauriejim** [comment]
Hello @stefba ! Thank you for reporting this.
By default, we are limiting the number of relations we populate by default.
You should receive the `info: ID` normally, I will confirm that with the tech team.
To be able to get more date, you will have to customize your controller function.

**Path —** `api/selection/controllers/selection.js`
```js
const { sanitizeEntity } = require('strapi-utils');

module.exports = {
  async find(ctx) {
    const entry = await strapi.services.selection.find(['entry', 'entry.piece', 'entry.piece.info']);
    return entry;
  },
};
```

**Comment by lauriejim** [comment]
*This message is to help for the tech team*

This is the response we have:

```json
{
    "id": 1,
    "created_at": "2020-04-15T14:27:49.890Z",
    "updated_at": "2020-04-15T14:27:49.890Z",
    "entry": [
        {
            "id": 1,
            "piece": {
                "id": 1,
                "title": "P1",
                "created_at": "2020-04-15T14:27:10.358Z",
                "updated_at": "2020-04-15T14:27:10.358Z"
            }
        },
        {
            "id": 2,
            "piece": {
                "id": 2,
                "title": "P2",
                "created_at": "2020-04-15T14:27:17.757Z",
                "updated_at": "2020-04-15T14:27:17.757Z"
            }
        }
    ]
}
```

I think the key `info` is missing. We should have it with the `ID` of the compo.
BUT we can't get a compo entry with it's `ID`
SO should we populate, in any cases, component values?

**Comment by s-barsch** [comment]
Thank you for you response!

I couldn’t make the controller function work. I already wrote a function that joins the two queries (`/pieces` and `/selection`), so I will maybe look into this later.

From a users perspective, it feels odd that a key doesn’t show up in the response at all, with no notification whatsoever. But I don’t know what complexities would come with it to populate the component values or even the id.

**Comment by lauriejim** [comment]
Can you try again on your side with the last version of Strapi?
and let me know if it works on your side.

**Comment by s-barsch** [comment]
Hey @lauriejim,

thanks for your response. (And sorry for my late response.)

I tested it again with a fresh Strapi installation – and unfortunately, my results are still the same! No info field.

Testing with **`3.0.0-beta.20.3`**, my results for `/selection` is:

```
{
   "id":1,
   "created_at":"2020-05-14T14:35:24.524Z",
   "updated_at":"2020-05-14T14:35:24.524Z",
   "entry":[
      {
         "id":1,
         "piece":{
            "id":1,
            "title":"first piece",
            "created_at":"2020-05-14T14:34:24.644Z",
            "updated_at":"2020-05-14T14:34:24.644Z"
         }
      },
      {
         "id":2,
         "piece":{
            "id":2,
            "title":"second piece",
            "created_at":"2020-05-14T14:34:35.103Z",
            "updated_at":"2020-05-14T14:34:35.103Z"
         }
      }
   ]
}
```

Whereas my `/pieces` response includes the info field:

```
[￼
   {￼
      "id":1,
      "title":"first piece",
      "created_at":"2020-05-14T14:34:24.644Z",
      "updated_at":"2020-05-14T14:34:24.644Z",
      "info":{
         "id":1,
         "Color":"red",
         "Material":"blue"
      }
   },
   {
      "id":2,
      "title":"second piece",
      "created_at":"2020-05-14T14:34:35.103Z",
      "updated_at":"2020-05-14T14:34:35.103Z",
      "info":{
         "id":2,
         "Color":"green",
         "Material":"green"
      }
   }
]
```

**Comment by lauriejim** [comment]
Can you please share you `api` & `components` folder.

**Comment by Oyinkansolight** [comment]
Wait this matter has not still been solved? 😢

**Reply by weidox** [reply]
Have encountered the same issue. Calling strapi.services.mycollection.find(query, populate) and findOne, mycollection does have a few array collection relations, of which only the newest added one gets property completely missing when it is not included into "populate" array. Others do work well, returning simple arrays of ID fields.
Strapi 3.1.5.

**Reply by weidox** [reply]
And I was lucky to determine what caused this issue, as this related collection was a bit different. Culprit is the relation type. When relation is one to many without backwards relation (dot-arrows image), it works. And when relation is one to many with backwards relation (dot-sticks image) it fails. That could be even intentional because when relation is with backwards link, it's still possible to find out related entities by doing search on related collection. And maybe there's difference in how relations are stored in such case, so maybe it's more costly to fetch plain ids when field is not populated. But in any case it is inconvenient and confusing.

**Reply by derrickmehaffy** [reply]
We are aware of the issue as it relates to: https://github.com/strapi/strapi/issues/7454

We will be working on a solution after engaging with our backend engineers

**Reply by weidox** [reply]
The #7454 seems to be a different issue, as it is about components, though it may be related. I found a matching closed issue #7361 which mentions open #4630 .

**Reply by derrickmehaffy** [reply]
Also related is: https://github.com/strapi/strapi/issues/7360

Functionally in the backend components and dynamic zones are handled almost exactly the same hence why they are related.
